lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -2415,11 +2415,15 @@ submitTransaction ctx apiw@(ApiT wid) apitx@(ApiSerialisedTransaction (ApiT seal
     let ourOuts = getOurOuts apiDecoded
     let ourInps = getOurInps apiDecoded
 
-    let allInpsNum = length $ apiDecoded ^. #inputs
+    let inpsWits = length $ L.nubBy samePaymentKey $
+            (apiDecoded ^. #inputs) ++ (apiDecoded ^. #collateral)
+    let wdrlWits = length $ apiDecoded ^. #withdrawals
+    let certWits = length $ filter certWitNeeded $ apiDecoded ^. #certificates
```

Comment:
This appears to ignore any potential for overlap of required witnesses between all of `inpsWits`, `wdrlWits` `certWits` (aside from among the inputs themselves).

So I believe this will overestimate the required number of witnesses if either:
1. A withdrawal key is identical to a input payment key
2. A certificate key is identical to a input payment key
3. There are multiple certs with the same key
4. There's a cert and withdrawal with the same key
5. Multiple withdrawals with the same key (absurd, not sure if technically possible)

I think (4) could be a common occurrence: users re-delegating and the wallet automatically withdrawing their rewards at the same time.

Reply:
Locally when I run `cabal v2-test cardano-wallet:integration --test-option=-j --test-option="10" --test-option="--match" --test-option "TRANS_NEW"` I do get a failure 
<img width="1376" alt="Skärmavbild 2022-02-08 kl  13 20 05" src="https://user-images.githubusercontent.com/304423/152985771-309da10b-2d06-44f3-83f0-68ed8179d797.png">

because the new tx workflow test for joining does [wait for rewards](https://github.com/input-output-hk/cardano-wallet/blob/baf862039942c05ab0b5e78874f66df51b742a2e/lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs#L2258-L2282) before rejoining.

In CI it appears to have passed because of the retrying `it` 😰

Reply:
Hm… never mind, I spoke too quickly. There are some Plutus related failures in CI however, and I guess the `retry` error logs are related to them.

Reply:
Looking into why then [this waiting for rewards](https://github.com/input-output-hk/cardano-wallet/blob/baf862039942c05ab0b5e78874f66df51b742a2e/lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs#L2258-L2282) still works with this PR, plus trying to change the wallet type to a `rewardWallet` for extra rewards in the beginning, it seems that the `expectField #withdrawals (`shouldNotBe` [])` fails for the first two joins… I.e. there are no actual withdrawals 🤔 (I thought there would be)

Reply:
it looks like there are needed 2 pools (as in hte test we are rejoining), and sometimes the second pool is not available

Reply:
I pushed an integration test demonstrating (4)

```
 src/Test/Integration/Framework/DSL.hs:2446:7:
  1) API Specifications, NEW_SHELLEY_TRANSACTIONS, TRANS_NEW_JOIN_01e - Can re-join and withdraw at once
       expected a successful response but got an error: ClientError (Object (fromList [("code",String "missing_witnesses_in_transaction"),("message",String "The transaction has 3 inputs and 2 witnesses included. Submit fully-signed transaction.")]))
       While verifying value:
         ( Status
             { statusCode = 403
             , statusMessage = "Forbidden"
             }
         , Left
             ( ClientError
                 ( Object
                     ( fromList
                         [
                             ( "code"
                             , String "missing_witnesses_in_transaction"
                             )
                         ,
                             ( "message"
                             , String "The transaction has 3 inputs and 2 witnesses included. Submit fully-signed transaction."
                             )
                         ]
                     )
                 )
             )
         )
```
and a simple fix: let's not count certs and withdrawals towards the number of required witnesses for now.